File: 130_ImplementTrie.py
"""
Problem: Implement Trie (Prefix Tree)
Description: Implement a trie with insert, search, and startsWith operations.

Approach: Each node contains 26 children (for 'a'-'z') and a flag for end of word.
"""

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True
    
    def search(self, word: str) -> bool:
        node = self.root
        for c in word:
            if c not in node.children:
                return False
            node = node.children[c]
        return node.isEnd
    
    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for c in prefix:
            if c not in node.children:
                return False
            node = node.children[c]
        return True

def main():
    trie = Trie()
    
    trie.insert("apple")
    print(trie.search("apple"))    # Output: True
    print(trie.search("app"))      # Output: False
    print(trie.startsWith("app"))  # Output: True
    trie.insert("app")
    print(trie.search("app"))      # Output: True

if __name__ == "__main__":
    main()

"""
Time Complexity: O(m) for all operations where m is length of word
Space Complexity: O(alphabet_size * n * m) where n is number of words
"""

File: 130_ImplementTrie.cpp
/*
Problem: Implement Trie (Prefix Tree)
Description: Implement a trie with insert, search, and startsWith operations.

Approach: Each node contains 26 children (for 'a'-'z') and a flag for end of word.
*/

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEnd;
    
    TrieNode() {
        for (int i = 0; i < 26; i++)
            children[i] = nullptr;
        isEnd = false;
    }
};

class Trie {
    TrieNode* root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                node->children[idx] = new TrieNode();
            node = node->children[idx];
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                return false;
            node = node->children[idx];
        }
        return node->isEnd;
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            int idx = c - 'a';
            if (!node->children[idx])
                return false;
            node = node->children[idx];
        }
        return true;
    }
};

int main() {
    Trie trie;
    
    trie.insert("apple");
    cout << trie.search("apple") << endl;    // Output: 1 (true)
    cout << trie.search("app") << endl;      // Output: 0 (false)
    cout << trie.startsWith("app") << endl;  // Output: 1 (true)
    trie.insert("app");
    cout << trie.search("app") << endl;      // Output: 1 (true)
    
    return 0;
}

/*
Time Complexity: O(m) for all operations where m is length of word
Space Complexity: O(alphabet_size * n * m) where n is number of words
*/
